### ISSUE TITLE
Auto-commit an application upon the first WSGI request

### ISSUE THREAD

**Issue by taschini** [morepath/morepath#392]
Currently, if you forget to explicitly commit an app before you start to serve it – quite a common occurrence if you are just starting out with morepath – you get an error.  Up to now the possible enhancements that have been taken into consideration were the following:

- A [check followed by an explanatory ConfigError](https://github.com/morepath/morepath/commit/c809ca84675ed1f17cd3434859adfb9c2a32555a) was attempted and then rolled back as it was breaking apps.
- In #378 we discussed the possibility of automatically committing the app on instantiation, and ultimately rejected it.

Now that #386 makes it possible, a further possibility is to have something like

```
if not self.is_committed():
    self.commit()
```

at the time of the first WSGI request.

This would improve usability without breaking any existing code: an app that has already been committed is left untouched, but morepath takes care of committing it for you in case you forgot.

Ultimately, this would lower the barrier for people transitioning from other micro-frameworks à la Flask.

**Comment by faassen** [comment]
We could probably do this in the publish method, which already only gets executed once for the very first request.

**Comment by faassen** [comment]
Could you try this out and see whether anything unexpected happens? 

If everything works we come up with a strategy to see how we can describe this in the docs.

**Comment by faassen** [comment]
This is a great improvement for just a single line of code added! 

Cleans up the tests quite a bit too. I wonder whether we should change our best practices recommendations in the documentation now to something like this:

* use autoscan to automatically scan extension code (and your own code). This is still useful to pick up morepath extensions automatically.

* no commit at all, no autosetup.

* then just immediately start serving.

If this works well we could eventually even deprecate autosetup and autocommit, though it won't hurt to use it and currently the query tool still requires this commit to be done so that's an idea for the future.

To summarize: code that previously used commit can now work entirely without commit, even if you mount apps and such.

@href, @henri-hulski, what do you think?

**Comment by href** [comment]
I like the change! But I'm not sure if I understand the implications for the docs yet. We mostly seem to be advertising the following pattern:

```python
    morepath.autosetup()
    morepath.run(App())
```

Since autoscan is invoked by autosetup and since it is still needed (to detect extensions), is there anything we can reasonably change? Wouldn't we have to wait until autoscan becomes obsolete as well?

Where the call is not necessary anymore I'm all for removing it though.

**Comment by henri-hulski** [comment]
If we get this to work without side effects, it really simplify the setup and I think also the tests. 
So how a Morepath setup would look like after the change?

**Comment by faassen** [comment]
@href, true, the setup wouldn't change all that much, but we could simplify it do use:

>  morepath.autoscan()
>  morepath.run(App())

i.e. autoscan only auto-scans packages and does not do an autocommit in addition.

**Comment by taschini** [comment]
I think point 6 of the code walkthrough in the quick start is probably one of the first topic that would need to be changed.

I do think that autocommit and autosetup should be phased out, the former to be made redundant and the latter to be replaced by autoscan.

**Comment by faassen** [comment]
There are valid use cases for committing explicitly before the first request -- debugging and performance measurement for instance. But those could be handled by ``App.commit()`` -- only a single app is going to be the WSGI app and all mounted apps will be recursively committed in this case. So we would still have an explicit commit. 

The only other thing that's needed is automatically importing packages that Morepath depends on so the directives get registered, that is, autoscan.

So I think @taschini is right and we could phase out autocommit and autosetup. We could actually make autosetup become identical to what autoscan is now -- I don't think it any behavior will be broken, the commit will just shift from autosetup to the first request. Phasing out autocommit would let us drop a hard-to-test global from Dectate.

**Comment by faassen** [comment]
Closing this one as discussion can continue elsewhere now.

**Comment by faassen** [comment]
I'm having second thoughts about automatically committing during the first request. The problem with it is that if there are any configuration conflicts you need to actually issue a request with a web browser in order to find them. The "sanity check" of just starting the application does nothing. I find myself writing in the docs "well, you can omit the commit but if you do so you can't really see issues before the first request so we recommend you commit anyway" and this makes me wonder what's the point.

**Comment by faassen** [comment]
It seems to me that while instantiation is commit is more likely to break an app it might be an issue we want to deal with instead. It doesn't have this issue, which I'm sorry I forgot about until I started having to write about the changes.

**Comment by taschini** [comment]
> The problem with it is that if there are any configuration conflicts you need to actually issue a request with a web browser in order to find them.

This is not different from any other potential mistakes in the code: you have to run it in order to see if it works or not.  How do you know if a certain view will trigger an Error 500? You issue a request.

Do you want to do it manually? No, of course, you want to have a test-suite that tries to look for these problems. A configuration mistake is something that a test suite is guaranteed to catch.

So, what should our recommendation be: explicitely invoke commit or have a test suite in place?

But, you might say, what about *exploratory development*, trying out things? Well in that case you probably already have the browser open anyway.  In this use case, you should be probably using something like the werkzeug server with autoreload and debugging on, instead of ``morepath.run``.

**Comment by faassen** [comment]
As an experiment I removed the commit during first request and replaced it with this on ``App``:

```
    def __new__(cls, *args, **kw):
        if not cls.is_committed():
            cls.commit()
        return super(App, cls).__new__(cls, *args, **kw)
```

All the tests still pass when I do this. Bizarrely enough even ``test_instantiation_before_config``, which has me scratching my head.

**Comment by faassen** [comment]
Oh, sorry, I was testing with the wrong test suite, ``test_instantiation_before_config`` fails as expected. Phew.

**Comment by faassen** [comment]
Unfortunately the above general ``__new__`` does not work with Python 3. I'm currently puzzling how to make it work. I am trying to avoid a contract where subclasses of ``App`` need to do a super just to get the commit working.

**Comment by faassen** [comment]
Ah, this works:

```
    def __new__(cls, *args, **kw):
        if not cls.is_committed():
            cls.commit()
        return super(App, cls).__new__(cls)
```

**Comment by faassen** [comment]
I'll leave the commit it in first request for the time being. Here are some tradeoffs:

* commit during first request means configuration errors only show up after first request. Also slows first request down.

* doing commit upon instantiation can be made to work, but it would break existing code that instantiates during import-time in top level of a module. This is done if you want to wrap middleware around the morepath app and your middleware is a top-level function.

* doing commit on request allows a pattern where a simple WSGI function can be constructed on top-level by instantiating without side-effects. Note though that there seems to be no problem sneaking in an explicit commit or instantiation into real-world WSGI server setups, see #205 

It could be we want to revisit this into the future, but re-closing this for now.
